Armv7 build neon

kwizart opened the pull request:
## Description
Building on arm fails when the builder lacks neon and/or neon isn't enforced with the kodi build.

## Motivation and Context
When building for armv7-a, neon fpu is optional. Altought it can be easier to enable it everywhere, there are several hardware that lack neon support, but still would be relevant for kodi usage.
One case among others is tegra20 (such as Trimslice) where the grate-driver project allow to use the tegra_vde ip block to provide video acceleration using the vdpau API.

## How Has This Been Tested?
Tested at build time with gcc using native compilation on the distro build infra.
Runtime tests are under progress using hw with/without neon support.

## Screenshots (if appropriate):
Build.log from the RPM Fusion infra for fedora 28.
https://koji.fedoraproject.org/koji/taskinfo?taskID=28720079

## Types of change
- [] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

hudokkow commented on xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt:
```diff
@@ -56,6 +56,7 @@ endif()
 if(ARCH MATCHES arm)
   list(APPEND SOURCES yuv2rgb.neon.S)
   list(APPEND HEADERS yuv2rgb.neon.h)
+  set_source_files_properties(yuv2rgb.neon.S PROPERTIES COMPILE_FLAGS -mfpu=neon)
```
`COMPILE_FLAGS` is deprecated. See https://cmake.org/cmake/help/latest/prop_tgt/COMPILE_FLAGS.html

hudokkow replied:
And congrats on your first contribution!

kwizart replied:
Thx for the review, I will change for COMPILE_DEFINITIONS

FernetMenta commented on xbmc/rendering/MatrixGL.cpp:
```diff
@@ -129,6 +129,8 @@ void CMatrixGL::Rotatef(GLfloat angle, GLfloat x, GLfloat y, GLfloat z)
 }
 
 #if defined(HAS_NEON) && !defined(__LP64__)
+#pragma GCC push_options
```
please refrain from thinks like this. We are in the process of getting rid of ifdefs and push_option is even more hacky.
move platform code to a separate file and include via cmake

wsnipex replied:
will also break on non-gcc toolchains

kwizart replied:
Thx for the reviews
I'm not sure I will be able to turn this file as pure asm .S file which should probably be preferable to a separate file with only the asm volatile and Matrix4Mul function as as cpp.

kwizart commented:
About vdpau in the tegra20 case, I'm not sure the v4l2 hw codec driver will be ready in time.
Now maybe there will be a bridge between API (there is already a v4l2 request api/ VA-API bridge by bootlin:
https://github.com/bootlin/libva-v4l2-request

But this PR is only about Neon. 
If neon can be kept optional in kodi on armv7-a it would allows some corner cases devices to be hacked to multimedia center. 

As others devices lacking neon are the Marvell armada devices (xp/370->385) 
This one is used in the scaleway-c1 "physical cloud" (that most Linux distro use as native builders).

FYI, I also have production devices of theses (originally used a light client devices from manufactured from Wyse).

Of course as the PI already has neon support since version 2, any other devices are corner cases anyway.

kwizart committed:
Move Matrix4Mul into a separate file to use -fpu=neon on it

v13: Assume ARCH MATCHES arm is always 32bit
v12: add ENABLE_NEON and not NEON_CFLAGS conditions
v8: Uses ARCH MATCHES arm
v2: move -fpu=neon to gl/gles sub-directories

Signed-off-by: Nicolas Chauvet <kwizart@gmail.com>

kwizart commented:
Hi there,

I've updated this PR adding a check for gcc before using pragma gcc in order to avoid any issue with others compilers.

I've tried to split the neon specific asm code into a separate file (look at kwizart/armv7_build_neon_v3 - but it's not working yet). This looks a very intrusive change.

Using #pragma gcc seems more on purpose. So right now I don't understand the concern around dropping ifdef and others things like this ? @FernetMenta Can you elaborate ?

Also to note that I'm not sure all aarch64 SOCs (where neon asm is the default but seems rather different) that has 32bit capacity can work with armv7-a neon. So a 32bit guest on aarch64 might not expose neon capacity. (I have a bug on a Jetson-TX1 with neon at least).

Thx for review.

FernetMenta commented:
> Can you elaborate

We have already too much ifdef magic in generic code. It is much better to add platform specific code into separate files and add them with cmake.

Why don't you remove the entire neon code from MatrixGL.cpp and add it somewhere else. Use cmake to add it if neon is enabled.

kwizart commented:
FYI, I' currently testing an updated patch on my side.
Hopefully the ci will not find an issue on its side.

Rechi commented on xbmc/rendering/MatrixGL.h:
```diff
@@ -83,3 +83,7 @@ class CMatrixGLStack
+
+#if defined(HAS_NEON) && !defined(__LP64__)
+void Matrix4Mul(float* src_mat_1, const float* src_mat_2);
+#endif
```
move this to `MatrixGL.cpp`

Rechi commented on xbmc/rendering/gl/CMakeLists.txt:
```diff
@@ -8,4 +8,11 @@ set(HEADERS GUIWindowTestPatternGL.h
             GLShader.h)
 
+if(ARCH MATCHES arm)
+  if(HAVE_NEON)
```
not defined anywhere

Rechi commented on xbmc/rendering/gles/CMakeLists.txt:
```diff
@@ -7,5 +7,12 @@ if(OPENGLES_FOUND)
               GLESShader.h)
 
+  if(ARCH MATCHES arm)
+    if(HAVE_NEON)
```
not defined anywhere

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/CMakeLists.txt:
```diff
@@ -56,6 +56,7 @@ endif()
 if(ARCH MATCHES arm)
```
`if(ARCH MATCHES arm AND ENABLE_NEON)`

Rechi commented on xbmc/rendering/MatrixGL.neon.cpp:
```diff
@@ -0,0 +1,49 @@
+ */
+
+
+#if defined(HAS_NEON) && !defined(__LP64__)
```
remove

kwizart replied:
I've tried to remove it, but it failed, this code is triggered on IOS-ARM64 despite the cmake condition. cmake seems weak to detect host versus target capabilities.
The other yuv2rgb.neon.S seems to work fine on aarch64.

kwizart replied:
I will try to rework the cmake condition anyway.

Rechi commented on xbmc/rendering/gl/CMakeLists.txt:
```diff
@@ -8,4 +8,9 @@ set(HEADERS GUIWindowTestPatternGL.h
             ../MatrixGL.h
             GLShader.h)
 
+if(ARCH MATCHES arm)
```
`if(ARCH MATCHES arm AND ENABLE_NEON)`

Rechi replied:
If `ENABLE_NEON` isn't true the C(++) define `HAS_NEON` will also not be set. If that isn't set Matrix4Mul will never be called.

https://github.com/xbmc/xbmc/blob/e97f19f0bd0987224e37a0ad70f76362ec71f611/cmake/scripts/common/ArchSetup.cmake#L147-L149
https://github.com/xbmc/xbmc/blob/4398f988e1577495ab6a9ba1e43e6bd2ac65f465/xbmc/rendering/MatrixGL.cpp#L174-L177

kwizart replied:
I confirm that I have -DHAS_NEON in my environment, but by chance, the NEON_FLAGS isn't defined, so I can set the one appropriate for my build. (that I expect will work for any build but might be less optimized than some others options).
So right now I'm (testing) adding -mfpu=neon unless NEON_FLAGS is already set.

Rechi commented on xbmc/rendering/gles/CMakeLists.txt:
```diff
@@ -7,5 +7,10 @@ if(OPENGLES_FOUND)
               ../MatrixGL.h
               GLESShader.h)
 
+  if(ARCH MATCHES arm)
```
`if(ARCH MATCHES arm AND ENABLE_NEON)`

kwizart commented:
I wonder if the way forward wouldn't be to remove the MatrixGL implementation and use glm instead ?
https://glm.g-truc.net/

Rechi commented on xbmc/rendering/gl/CMakeLists.txt:
```diff
@@ -8,4 +8,11 @@ set(HEADERS GUIWindowTestPatternGL.h
             GLShader.h)
 
+if(ARCH MATCHES arm AND ENABLE_NEON)
+  if (CMAKE_SIZEOF_VOID_P EQUAL 4)
```
arm is always 32 bit

Rechi commented on xbmc/rendering/gles/CMakeLists.txt:
```diff
@@ -7,5 +7,14 @@ if(OPENGLES_FOUND)
               GLESShader.h)
 
+  if(ARCH MATCHES arm AND ENABLE_NEON)
+    if(CMAKE_SIZEOF_VOID_P EQUAL 4)
```
arm is always 32 bit
